Core/src/IfcGeometryConverter/FaceConverter.h

```diff
@@ -768,132 +767,165 @@ namespace OpenInfraPlatform {
 				{
 					if (face.expired()) {
 						throw oip::ReferenceExpiredException(face);
 					}
-					// Indicates if conversion has failed
-					bool conversionFailed = false;
 
-					// Id of face in step file
-					const uint32_t faceID = face->getId();
+					// get attribute 1:
+					// list of bound loops (IfcFaceBound) with outer boundary loop (IfcFaceOuterBound) at index 0;
+					// one bound loop contains a list of loop points (carve::gem::vector<3>)
+					std::vector<std::vector<carve::geom::vector<3>>> faceBoundLoops;
+					convertIfcFaceBoundList(face->Bounds, pos, faceBoundLoops);
 
-					// To triangulate the mesh, carve needs 2D polygons, we collect the data in 2D and 3D for every bound
-					std::vector<std::vector<carve::geom2d::P2>> faceVertices2D;
-					std::vector<std::vector<carve::geom::vector<3>>> faceVertices3D;
+					// if simple case: one triangle without inner bound (= hole)
+					//  -> simple direct triangle construction is possible (probably fast)
+					if ((faceBoundLoops.size() == 1) && (faceBoundLoops[0].size() == 3))
+					{
+						// std::vector<carve::geom::vector<3>> loopVertices3D = faceBoundLoops[0];
+						addTriangleToPolyhedronData(faceBoundLoops[0], polygon, polygonIndices);
+					}
+					else
+					{
+						// general case: arbitrary number of vertices, possible inner bound (= hole)
+						//  -> elaborate triangulation with respect to arbitrary number of vertices and holes is necessary;
 
-					// Save polygon indices of merged vertices
-					std::map<uint32_t, uint32_t> mergedIndices;
-					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> modBounds;
-					modBounds.reserve(2);
-					bool faceLoopReversed = false;
+						addArbitraryFaceToPolyhedronData(face, faceBoundLoops, polygon, polygonIndices);
+					}
+				}
+				
+				void convertIfcFaceBoundList(
+					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> ifcFaceBounds,
+					const carve::math::Matrix& pos,
+					std::vector<std::vector<carve::geom::vector<3>>>& faceBoundLoops) const noexcept(true)
+				{
+					// As carve expects outer boundary of face to be at first index, outer boundary has to be moved to index 0 and inner boundary has index >= 1
+					for (auto it = ifcFaceBounds.begin(); it != ifcFaceBounds.end(); it++)
+					{
+						if (it->isOfType<typename IfcEntityTypesT::IfcFaceOuterBound>())
+						{
+							// swap element 'it' with first element
+							std::rotate(ifcFaceBounds.begin(), it, it + 1);
 
-					// Loop through all boundary definitions
-					int boundID = -1;
+							// according to ifc-documentation, only one boundary (= loop) can be outer boundary:
+							// "One loop is optionally distinguished as the outer loop of the face."
+							// (https://standards.buildingsmart.org/IFC/DEV/IFC4_3/RC3/HTML/link/ifcface.htm)
+							// thus, break the for-loop after THE outer boundary was found
+							break;
+						}
+					}

```

Comment:
I propose to make this a distinct function.

Takes in `std::vector<EXPRESSReference<... IfcFaceBound>>& faceBound` and puts the outer bound at index 0.

This way, the input to `convertIfcFaceBoundList`  can be made const reference (`const ...&`).

Reply:
Done in commit b4c24cb.
However, the input of the new distinct function `swapOuterBoundaryToFront` can't be a reference to a constant vector because the function `std::rotate(...)` performs changes on this vector.

An alternative approach to avoid the operation on the vector (and prevent the copy with the swap of elements) could be to only indicate the index of the outer boundary. In the function `convertIfcFaceBoundList`, the outer boundary would be added as first element to the target `faceBoundLoops`, and then the for-loop would add all other loops (with skipping the outer boundary).
At least in terms of comprehensible code, probably this would be more complex.

Reply:
not const is perfect, simple code is preferred :)